ggz-client-libs/man/ggz-wrapper.6
.TH "ggz-wrapper" "6" "0.0.14" "The GGZ Development Team" "GGZ Gaming Zone"
.SH "NAME"
.LP
ggz\-wrapper \- GGZ Gaming Zone command line core client
.SH "SYNOPSIS"
.LP
\fBggz\-wrapper\fR [\fB\-\-gametype\fR \fIGAMENAME\fP] [\fB\-\-user\fR \fINAME\fP]
[\fB\-\-destnick\fR \fINAME\fP] [\fIOPTION\fR...]
.SH "DESCRIPTION"
.LP
The GGZ Gaming Zone allows people to play many games online using so-called
GGZ core clients which are chat clients with integrated game launchers.
.LP
The \fBggz-wrapper\fR tool is a small special GGZ core client for the command
line. It does not support chat or display of rooms, tables and players,
but instead is useful for fast launching of online games.
Therefore, its main purpose is to be invoked from instant messengers such
as \fBgaim\fR or \fBkopete\fR.
.LP
It currently supports all two-player games, as human against human or
human against bot. If no options are given, \fBggz-wrapper\fR will invoke a
Tic-Tac-Toe game client on the default GGZ server, \fIlive.ggzgamingzone.org\fR,
logging in as guest user.
.SH "OPTIONS"
.TP
\fB\-\-gametype=\fR\fIGAMENAME\fP
The name of the game to play, e.g. TicTacToe or Chess.
This option expects a case-sensitive argument which must match one of the
available game modules in \fIggz.modules\fR. If the game type is not supported
by the specified GGZ server, the invocation will fail.
.TP
\fB\-\-user=\fR\fINAME\fP
The name of the player who is running the client. The default is to run with
an automatically generated guest user name. In order to use a registered name,
the option --password is implied.
.TP
\fB\-\-destnick=\fR\fINAME\fP
The name of the other player. If omitted or empty, the player running the client
will launch the table (game host). If empty, other player will be the bot.
If not empty, player will join the table which the other player already launched.
.TP
\fB\-\-password=\fR\fIPASSWORD\fP
Password to use in case of playing as a registered user. Note that on a
multi-user systems, other users might see the password, which is a security risk.
.TP
\fB\-\-server=\fR\fISERVER\fP
The server to play on. If not specified, the default server is
\fIggz://live.ggzgamingzone.org\fR.
.TP
\fB\-\-port=\fR\fIPORT\fP
The port to use to connect to the GGZ server. \fIPORT\fR is almost always
5688, so this option is rarely needed.
.LP
.SH "EXAMPLES"
.LP
To play Reversi against a bot, the invocation should be of the form
\fIggz-wrapper -s localhost -g Reversi -u P1 -f gtk -d ""\fR.
.SH "AUTHORS"
.LP
The GGZ Development Team
<ggz\-dev@mail.ggzgamingzone.org>
.SH "SEE ALSO"
.LP 
ggz.modules(5), ggz(7)


ggz-client-libs/ggz-wrapper/ggz-wrapper.c
/*
 * File: ggz-wrapper.c
 * Author: Josef Spillner
 * Project: GGZ Client libs
 * Date: 2004
 * $Id: ggz-wrapper.c 9165 2007-07-04 13:58:17Z josef $
 *
 * Code for a wrapper for GGZ games
 *
 * Copyright (C) 2004-2005 GGZ Development Team
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>			/* Site-specific config */
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>

#include "server.h"
#include "loop.h"
#include "game.h"

#define DEFAULT_SERVER   "live.ggzgamingzone.org"
#define DEFAULT_PORT     5688
#define DEFAULT_GAMENAME "TicTacToe"
#define DEFAULT_NICKNAME NULL
#define DEFAULT_PASSWORD NULL
#define DEFAULT_DESTNICK NULL

#define _(x) x

char *game_name;
char *dst_nick;
char *frontend;

static void usage(void) {
	printf(_("ggz-wrapper: Dummy core client for the GGZ Gaming Zone\n"));
	printf(_("\n"));
	printf(_("Options:\n"));
	printf(_("[-g <gametype>] Type of game to play\n"));
	printf(_("[-u <user>    ] Your nickname\n"));
	printf(_("[-p <password>] Password (empty for guest logins)\n"));
	printf(_("[-s <server>  ] Server hostname\n"));
	printf(_("[-P <port>    ] Server port\n"));
	printf(_("[-d <destnick>] Nickname of opponent\n"));
	printf(_("[-f <frontend>] Preferred game client frontend\n"));
	exit(0);
}

int main(int argc, char **argv) {
	struct option options[] =
	{
		{"help", no_argument, 0, 'h'},
		{"gametype", required_argument, 0, 'g'},
		{"user", required_argument, 0, 'u'},
		{"password", required_argument, 0, 'p'},
		{"server", required_argument, 0, 's'},
		{"port", required_argument, 0, 'P'},
		{"destnick", required_argument, 0, 'd'},
		{"frontend", required_argument, 0, 'f'},
		{0, 0, 0, 0}
	};
	char optstring[] = "g:u:p:s:d:P:f:h";
	char randomuser[64];

	int optch;
	int optindex;

	char *server = DEFAULT_SERVER;
	char *password = DEFAULT_PASSWORD;
	char *nick = DEFAULT_NICKNAME;
	int port = DEFAULT_PORT;

	game_name = DEFAULT_GAMENAME;
	dst_nick = DEFAULT_DESTNICK;
	frontend = NULL;

	while((optch = getopt_long(argc, argv, optstring, options, &optindex)) != -1) {
		switch(optch) {
			case 'g':
				game_name = optarg;
				break;
			case 'u':
				nick = optarg;
				break;
			case 'p':
				password = optarg;
				break;
			case 's':
				server = optarg;
				break;
			case 'd':
				dst_nick = optarg;
				break;
			case 'P':
				port = atoi(optarg);
				break;
			case 'f':
				frontend = optarg;
				break;
			case 'h':
				usage();
				break;
			default:
				printf(_("Option %c unknown, try --help.\n"), optch);
				usage();
				break;
		}
	}

	if(!nick) {
		snprintf(randomuser, sizeof(randomuser), _("guest%i"), rand() % 10000);
		nick = strdup(randomuser);
	}

	if(password)
		server_init(server, port, GGZ_LOGIN, nick, password);
	else
		server_init(server, port, GGZ_LOGIN_GUEST, nick, password);
	loop();
	return 0;
}
